src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/DecompressionHandler.cs

```diff
@@ -224,12 +223,203 @@ public DeflateDecompressedContent(HttpContent originalContent)
+                    {
+                        if (_decompressionStream is null)
+                        {
+                            int firstByte = await _stream.PeekFirstByteAsync(cancellationToken).ConfigureAwait(false);
+                            _decompressionStream = CreateDecompressionStream(firstByte, _stream);
+                        }
+
+                        await _decompressionStream.CopyToAsync(destination, bufferSize, cancellationToken).ConfigureAwait(false);
+                    }
+                }
+
+                private static Stream CreateDecompressionStream(int firstByte, Stream stream) =>
+                    (firstByte & 0xF) == 8 ?
+                        new ZLibStream(stream, CompressionMode.Decompress) :
+                        new DeflateStream(stream, CompressionMode.Decompress);
+
+                private sealed class PeekFirstByteReadStream : HttpBaseStream
+                {
+                    private readonly Stream _stream;
+                    private byte _firstByte;
+                    private FirstByteStatus _firstByteStatus;
+
+                    public PeekFirstByteReadStream(Stream stream) => _stream = stream;
+
+                    protected override void Dispose(bool disposing)
+                    {
+                        if (disposing)
+                        {
+                            _stream.Dispose();
+                        }
+                        base.Dispose(disposing);
+                    }
+
+                    public override bool CanRead => true;
+                    public override bool CanWrite => false;
+                    public override ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken) => throw new NotSupportedException();
+
+                    public int PeekFirstByte()
+                    {
+                        Debug.Assert(_firstByteStatus == 0);
+
+                        int value = _stream.ReadByte();
+                        if (value == -1)
+                        {
+                            _firstByteStatus = FirstByteStatus.Consumed;
+                            return -1;
+                        }
+
+                        _firstByte = (byte)value;
+                        _firstByteStatus = FirstByteStatus.Available;
+                        return value;
+                    }
+
+                    public async ValueTask<int> PeekFirstByteAsync(CancellationToken cancellationToken)
+                    {
+                        Debug.Assert(_firstByteStatus == 0);
+
+                        var buffer = new byte[1];
+
+                        int bytesRead = await _stream.ReadAsync(buffer, cancellationToken).ConfigureAwait(false);
+                        if (bytesRead == 0)
+                        {
+                            _firstByteStatus = FirstByteStatus.Consumed;
+                            return 0;
```

Comment:
This behavior is different than the sync behavior above -- returning 0 vs -1. Intentional?

Reply:
No, that should be -1.